lib/maths/CBoostedTreeImpl.cc

```diff
@@ -423,13 +441,15 @@ CBoostedTreeImpl::crossValidateForest(core::CDataFrame& frame) {
         folds.pop_back();
         TNodeVecVec forest;
         double loss;
-        std::tie(forest, loss) = this->trainForest(
+        TDoubleVec lossValues;
+        std::tie(forest, loss, lossValues) = this->trainForest(
             frame, m_TrainingRowMasks[fold], m_TestingRowMasks[fold], m_TrainingProgress);
         LOG_TRACE(<< "fold = " << fold << " forest size = " << forest.size()
                   << " test set loss = " << loss);
         lossMoments.add(loss);
         m_FoldRoundTestLosses[fold][m_CurrentRound] = loss;
         numberTrees.push_back(static_cast<double>(forest.size()));
+        m_Instrumentation->lossValues(std::to_string(fold), std::move(lossValues));
```

Comment:
Is there any reason to convert this to a string here? I can't think of one.

Reply:
Yes, when I am going to write the loss values in rapid json, `fold` becomes the key and it has to be a string. I cannot convert it to string during writing since the string has to exist in memory until the writer finishes writing, otherwise, it just puts some random bytes as the key.